prune Bazel build files from imported go deps and update to gazelle 0.10.1

ixdy opened the pull request:
**What this PR does / why we need it**: updates the `hack/godep-save.sh` script to prune any imported `BUILD`, `BUILD.bazel`, or `WORKSPACE` files.

We assume anything imported through godep doesn't need Bazel, but a Bazel-enabled Go project may include `BUILD` files with missing dependencies not imported by godep. To prevent this from breaking the Bazel build, remove these. (Gazelle will then regenerate the `BUILD` files based on the go build metadata.)
 
Additionally, there seems to be a bug in gazelle where it follows the `vendor/k8s.io` staging symlinks only when run through `hack/run-in-gopath.sh`. Updating to gazelle 0.10.1 fixes this bug.

A similar PR was recently merged in kubernetes/test-infra: https://github.com/kubernetes/test-infra/pull/7366

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

/assign @thockin 
cc @jayconrod

ixdy committed:
Remove all upstream BUILD, BUILD.bazel, and WORKSPACE files from vendor/

ixdy committed:
run hack/godep-restore.sh && hack/godep-save.sh

ixdy committed:
Update to gazelle 0.10.1

cblecker commented on Godeps/Godeps.json:
```diff
@@ -1976,6 +1976,7 @@
 		},
 		{
 			"ImportPath": "github.com/inconshreveable/mousetrap",
+			"Comment": "v1.0",
```
Interesting that it's adding these back in.

ixdy replied:
🤷‍♂️  who even understands godep?

cblecker replied:
:left-shark:

cblecker commented on hack/godep-save.sh:
```diff
@@ -81,6 +81,11 @@ done
+# Assume that anything imported through godep doesn't need Bazel to build.
+# Prune out any Bazel build files, since these can break the build due to
+# missing dependencies that aren't included by godep.
+find vendor/ -type f \( -name BUILD -o -name BUILD.bazel -o -name WORKSPACE \) \
```
Would this logic make more sense as part of `hack/update-bazel.sh`?

ixdy replied:
I thought about that, but I didn't want to unnecessarily recreate these files all the time, just only when we're actually updating `vendor/`

cblecker replied:
My thought is, if we're calling `hack/update-bazel.sh` for some reason, we are specifically saying we want to update the bazel files. Is there a large performance difference between updating without deleting, vs deleting?

I also wonder if there's some weird situation we could get into where a build file is updated in some other way, and starts failing the verify-bazel step because of the way it recreates in a temp folder.

I'm not stuck on it, but part of me thinks if you're we're gonna blow them away at all, just blow them all away every time.

ixdy replied:
the other reason not to do this is that `kazel` prints out every time it updates a file, which it'll do every time these are recreated, even though there will likely be no diff.

BenTheElder commented:
/retest

ixdy commented:
/retest

thockin commented on hack/update-bazel.sh:
```diff
@@ -36,11 +36,16 @@ kube::util::go_install_from_commit \
+(
+  # gazelle needs to run from the real ${KUBE_ROOT}, as otherwise its symlink
+  # following logic can get confused.
+  cd "${KUBE_ROOT}"
```
Not sure I understand this - under what circumstances is it not?

ixdy replied:
recent functionality in gazelle causes it to follow symlinks, but only if they're outside the repo root.

if you run `hack/run-in-gopath.sh hack/update-bazel.sh`, gazelle ends up running inside `_output/local/go/src/k8s.io/kubernetes`. gazelle doesn't currently resolve the repo root to an absolute path, so its [check to verify that the symlink is out-of-repo](https://github.com/bazelbuild/bazel-gazelle/blob/257c283f6dd0d177cad931f3e6897f07fed5b41e/internal/packages/walk.go#L452) fails (since it resolves symlinks its considering to their absolute path).

For example, consider `vendor/k8s.io/apimachinery`, which points to `../../staging/src/k8s.io/apimachinery`.

When run in the real `$KUBE_ROOT`, gazelle detects the repo root as `$HOME/go/src/k8s.io/kubernetes` and the endpoint of this symlink as `$HOME/go/src/k8s.io/kubernetes/staging/src/k8s.io/apimachinery`. Since that's within the repo, it doesn't follow the symlink.

When run through `hack/run-in-gopath.sh`, gazelle detects the repo root as `$HOME/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes`, but still resolves the endpoint of the vendor symlink as `$HOME/go/src/k8s.io/kubernetes/staging/src/k8s.io/apimachinery`. It considers this as out-of-repo and thus follows the symlink, erroneously.

I tried passing `-repo_root ${KUBE_ROOT}` instead of the chdir, but this was causing different problems in vendor/ that I haven't yet figured out.

I also have refrained from changing the cwd for `kazel`, since `kazel` will blow up if it's not run from a valid GOPATH.

thockin commented:
LGTM, just one Q

cblecker commented:
/retest

BenTheElder commented:
/retest

cblecker commented:
/lgtm
/approve

thockin commented:
> recent functionality in gazelle causes it to follow symlinks, but only if
they're outside the repo root.

How does it calculate repo root?

> if you run hack/run-in-gopath.sh hack/update-bazel.sh, gazelle ends up
running inside _output/local/go/src/k8s.io/kubernetes. gazelle doesn't
currently resolve the repo root to an absolute path, so its check to verify
that the symlink is out-of-repo fails (since it resolves symlinks its
considering to their absolute path).

This sounds incredibly broken and fragile.  If it needs to do anything like
this, it should probably canonicalize the path, resolve symlinks, etc.
  `pwd -P`

> For example, consider vendor/k8s.io/apimachinery, which points to
../../staging/src/k8s.io/apimachinery.

> When run in the real $KUBE_ROOT, gazelle detects the repo root as
$HOME/go/src/k8s.io/kubernetes and the endpoint of this symlink as
$HOME/go/src/k8s.io/kubernetes/staging/src/k8s.io/apimachinery. Since
that's within the repo, it doesn't follow the symlink.

> When run through hack/run-in-gopath.sh, gazelle detects the repo root as
$HOME/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes, but
still resolves the endpoint of the vendor symlink as $HOME/go/src/
k8s.io/kubernetes/staging/src/k8s.io/apimachinery. It considers this as
out-of-repo and thus follows the symlink, erroneously.

Did you file a bug against gazelle?  This is broken.

thockin commented:
I'm happy of you just leave a TODO with a bug link to a bug against gazelle.


On Fri, Mar 23, 2018 at 3:02 PM Tim Hockin <thockin@google.com> wrote:

>
> > recent functionality in gazelle causes it to follow symlinks, but only if
> they're outside the repo root.
>
> How does it calculate repo root?
>
> > if you run hack/run-in-gopath.sh hack/update-bazel.sh, gazelle ends up
> running inside _output/local/go/src/k8s.io/kubernetes. gazelle doesn't
> currently resolve the repo root to an absolute path, so its check to verify
> that the symlink is out-of-repo fails (since it resolves symlinks its
> considering to their absolute path).
>
> This sounds incredibly broken and fragile.  If it needs to do anything like
> this, it should probably canonicalize the path, resolve symlinks, etc.
>   `pwd -P`
>
> > For example, consider vendor/k8s.io/apimachinery, which points to
> ../../staging/src/k8s.io/apimachinery.
>
> > When run in the real $KUBE_ROOT, gazelle detects the repo root as
> $HOME/go/src/k8s.io/kubernetes and the endpoint of this symlink as
> $HOME/go/src/k8s.io/kubernetes/staging/src/k8s.io/apimachinery. Since
> that's within the repo, it doesn't follow the symlink.
>
> > When run through hack/run-in-gopath.sh, gazelle detects the repo root as
> $HOME/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes, but
> still resolves the endpoint of the vendor symlink as $HOME/go/src/
> k8s.io/kubernetes/staging/src/k8s.io/apimachinery. It considers this as
> out-of-repo and thus follows the symlink, erroneously.
>
> Did you file a bug against gazelle?  This is broken.
>
>

ixdy commented:
I agree that it's fragile. I'm discussing this with @jayconrod out-of-band to confirm that it's a bug in gazelle (I'm pretty sure it is).

ixdy commented:
It looks like this was just fixed in https://github.com/bazelbuild/bazel-gazelle/pull/166.

Updating to a newer gazelle causes more churn in some of the build files due to https://github.com/bazelbuild/bazel-gazelle/pull/148, and it probably also requires updating rules_go.

I'm tempted to just leave a comment here pointing at https://github.com/bazelbuild/bazel-gazelle/pull/166 and then plan to update this soon and remove the workaround, rather than bumping everything now. WDYT @thockin?

cblecker commented:
/hold
for decision

jayconrod commented:
@ixdy I'll cherry-pick that fix back to Gazelle 0.10 and tag 0.10.1.

Gazelle at tip will require a newer version of rules_go. In particular, bazelbuild/bazel-gazelle#148 generates `importmap` attributes on `go_proto_library`, and those were recently introduced. We need to have `importmap` attributes for vendored libraries in order to avoid conflicts when multiple copies of the same package are linked (which is more common than you would think).

ixdy commented:
yeah, I was just about to comment that I tried updating to gazelle HEAD, and it complained that `RULES_GO_VERSION` is 0.10.0; updating to rules_go HEAD doesn't resolve that error message.

we're actually not using 0.10.0 right now, but rather 6 commits past it; we're using this version because we needed buildtools to be vendored.

ixdy commented:
though: this PR will unblock #57600, which will allow us to vendor gazelle and buildtools ourselves.

at that point, it won't matter whether the version of gazelle we're using vendors buildtools or not.

jayconrod commented:
In Gazelle 0.10.1, I'll include changes that vendor buildtools on our side, so you may not need to vendor it yourself unless you have a separate dependency.

jayconrod commented:
Gazelle [0.10.1](https://github.com/bazelbuild/bazel-gazelle/releases/tag/0.10.1) is ready.

ixdy commented:
/hold cancel

Updated to gazelle 0.10.1. Removed the chdir workaround; confirmed that we still get the same results via `hack/update-bazel.sh` and `hack/run-in-gopath.sh hack/update-bazel.sh`.

ixdy commented:
nope; it'll auto-detect it based on the presence of the `WORKSPACE` file.

cblecker commented:
/lgtm
